fastpickle.py

```diff
@@ -458,6 +459,9 @@ def __init__(self, file, protocol=None, *, fix_imports=True,
         self.write = self.framer.write
         self._write_large_bytes = self.framer.write_large_bytes
         self.memo = {}
+        self.threads_access = {} # maps id(obj) -> threads which accessed obj in the order accessed
+        self.local_memo = {} # for each thread, stores the number of items memoized in that thread.
+        self.r_memo = {}  # reverse memo, maps idx of memo -> thread_num, local_memo[thread_num]
```

Comment:
We can track `memo -> (offset_l: int, offset_r: int)` here and in `save`/`memoize`/etc., instead of backtracking the opcodes to find the byte range associated with a memo ID.